### PR TITLE
internal/docstore/dynamodocstore: add unordered actions

### PR DESCRIPTION
**Pull request by shantuo**
We use Transaction API for both read and write. The `splitActions` is different for unordered, it doesn't allow the same key to appear more than once through the entire action list. I think this should be a requirement at the driver+portable API level.

### CODE REVIEW

File: internal/docstore/drivertest/driverbenchmark.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
+// Copyright 2019 The Go Cloud Development Kit Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (jba) [inline internal/docstore/drivertest/driverbenchmark.go]: Should this file be in this PR?

File: internal/docstore/drivertest/drivertest.go [hunk 2]
```diff
@@ -161,7 +162,8 @@ func RunConformanceTests(t *testing.T, newHarness HarnessMaker, ct CodecTester,
 	t.Run("Data", func(t *testing.T) { withCollection(t, newHarness, testData) })
 	t.Run("TypeDrivenCodec", func(t *testing.T) { testTypeDrivenDecode(t, ct) })
 	t.Run("BlindCodec", func(t *testing.T) { testBlindDecode(t, ct) })
-	t.Run("MultipleActions", func(t *testing.T) { withCollection(t, newHarness, testMultipleActions) })
+	t.Run("OrderedActions", func(t *testing.T) { withCollection(t, newHarness, testOrderedActions) })
+	t.Run("UnorderedActions", func(t *testing.T) { withCollection(t, newHarness, testUnorderedActions) })
```
Comment (jba) [inline internal/docstore/drivertest/drivertest.go]: Is this from #1963? We can't submit that until I upgrade the gRPC replayer. I'm working on that (trying to update cloud.google.com/go to head).

File: internal/docstore/drivertest/drivertest.go [hunk 3]
```diff
@@ -815,7 +817,9 @@ func addQueryDocuments(t *testing.T, coll *ds.Collection) {
 }
 
 func testGetQuery(t *testing.T, coll *ds.Collection) {
-	cleanUpTable(t, newHighScore, coll)
+	if err := cleanUpTable(newHighScore, coll); err != nil {
```
Comment (jba) [inline internal/docstore/drivertest/drivertest.go]: I thought cleanUpTable takes a `t` and calls `t.Fatal` itself?

File: internal/docstore/drivertest/drivertest.go [hunk 4]
```diff
@@ -1080,8 +1086,54 @@ func testMultipleActions(t *testing.T, coll *ds.Collection) {
 	}
 }
 
+func testUnorderedActions(t *testing.T, coll *ds.Collection) {
+	if err := cleanUpTable(newDocmap, coll); err != nil {
```
Comment (jba) [inline internal/docstore/drivertest/drivertest.go]: This looks very close to testUnorderedActions from #1963, but I see some differences. I think we need to wait for that PR to go in.

File: internal/docstore/dynamodocstore/dynamo.go [hunk 5]
```diff
@@ -279,6 +283,75 @@ func (c *collection) primaryKey(a *driver.Action) [2]interface{} {
...
+	collect := func(cur *[]*driver.Action) { // called when the current group is known to be finished
+		if len(*cur) > 0 {
+			groups = append(groups, *cur)
+			*cur = nil
+		}
+	}
+	for _, a := range actions {
+		if keys := c.primaryKey(a); keys[0] != nil {
+			if km[keys] {
+				return nil, fmt.Errorf("repeated key: %v, %v; item must be unique for unordered actions", keys[0], keys[1])
```
Comment (jba) [inline internal/docstore/dynamodocstore/dynamo.go]: You can return the index of the bad action (the second one) as well.

File: internal/docstore/dynamodocstore/dynamo.go [hunk 6]
```diff
@@ -279,6 +283,75 @@ func (c *collection) primaryKey(a *driver.Action) [2]interface{} {
 	return keys
 }
 
+func (c *collection) runActionsUnordered(ctx context.Context, actions []*driver.Action, opts *driver.RunActionsOptions) driver.ActionListError {
+	errs := make([]error, len(actions))
+	groups, err := c.splitActionsUnordered(actions)
+	if err != nil {
+		errs[0] = err
```
Comment (jba) [inline internal/docstore/dynamodocstore/dynamo.go]: Use the index of the bad action here (see below).

File: internal/docstore/dynamodocstore/dynamo_test.go [hunk 7]
```diff
@@ -39,11 +39,18 @@ import (
 //   local index:  "Game" partition key, "Score" sort key
 //   global index: "Player" partition key, "Time" sort key
 // The conformance test queries should exercise all of these.
+//
```
Comment (jba) [inline internal/docstore/dynamodocstore/dynamo_test.go]: Remove this file from the PR too.

File: internal/docstore/drivertest/drivertest.go (hunk 3 above)
Comment (shantuo) [reply internal/docstore/drivertest/drivertest.go]: Yes, I wanted to use it with benchmarks as well. Moved to a separate PR.

File: internal/docstore/drivertest/drivertest.go (hunk 4 above)
Comment (shantuo) [reply internal/docstore/drivertest/drivertest.go]: Yes, removed.

File: internal/docstore/dynamodocstore/dynamo_test.go (hunk 7 above)
Comment (shantuo) [reply internal/docstore/dynamodocstore/dynamo_test.go]: Done.

File: internal/docstore/dynamodocstore/dynamo.go (hunk 5 above)
Comment (shantuo) [reply internal/docstore/dynamodocstore/dynamo.go]: Done.

File: internal/docstore/dynamodocstore/dynamo.go (hunk 6 above)
Comment (shantuo) [reply internal/docstore/dynamodocstore/dynamo.go]: Done.

File: internal/docstore/drivertest/drivertest.go (hunk 2 above)
Comment (jba) [reply internal/docstore/drivertest/drivertest.go]: I submitted #1963 by skipping the unordered test.

**Pull request merged**